# Nigerian Pidgin Sentiment Analyzer
## Data Validation and Exploration

Dataset: NaijaSenti Nigerian Pidgin Annotated Twitter Dataset

Objective:
- Understand dataset structure
- Identify data quality issues
- Validate labels
- Prepare data for preprocessing and modeling

In [2]:
import pandas as pd
import numpy as np

print("Libraries Imported Successfully")

Libraries Imported Successfully


In [3]:
train_path = "../dataset/raw/train.tsv"

train_df = pd.read_csv(
    train_path,
    sep="\t"
)

train_df.head()

,tweet,label
0,yeah ‍️the guy wants to trend dat was why e jo...,negative
1,this life is so funny sef you will work hard a...,negative
2,dis is unfair of urcompany goingtowks nw ive b...,negative
3,lil wayne don vex me im actually not excited a...,negative
4,dis is unfair of ur company going to wks nw iv...,negative


In [4]:
train_df.shape

(5121, 2)

In [5]:
train_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5121 entries, 0 to 5120
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   tweet   5121 non-null   str  
 1   label   5121 non-null   str  
dtypes: str(2)
memory usage: 80.1 KB


In [6]:
train_df['label'].value_counts()

label
negative    3241
positive    1808
neutral       72
Name: count, dtype: int64

In [7]:
# Checking Missing Data

train_df.isnull().sum()

tweet    0
label    0
dtype: int64

In [8]:
# Checking Duplicate Values

train_df.duplicated().sum()

np.int64(480)

## Duplicate Analysis

The training dataset contains 480 duplicate records.

Duplicates may introduce bias during model training because repeated samples can cause the model to memorize specific patterns instead of learning general sentiment representations.

The duplicates will be removed during the data preprocessing stage while preserving the original raw dataset.

In [9]:
# Investigating the duplicates

duplicate_rows = train_df[train_df.duplicated()]

duplicate_rows.head()

,tweet,label
1917,lol hiphop still livesabi cdq self no no wetin...,negative
1921,i don suffer o on top small twitter naim perso...,negative
1923,some people will see you drinking water and be...,negative
1941,this lingard wey una dey start hope say e no g...,negative
2002,see bah asides good health financial security ...,negative


In [10]:
duplicate_rows['label'].value_counts()

label
positive    240
negative    238
neutral       2
Name: count, dtype: int64

## Duplicate Records Analysis

A duplicate analysis identified 480 duplicate rows in the training dataset.

Duplicate sentiment distribution:

| Sentiment | Duplicate Count |
|-----------|----------------:|
| Positive  | 240 |
| Negative  | 238 |
| Neutral   | 2 |

Since duplicate records can introduce bias during model training by over-representing certain samples, they will be removed during the preprocessing phase. The original dataset will remain unchanged in the `dataset/raw` directory to preserve data integrity.

In [11]:
# Verifying Whether the Dataset Is Balanced

train_df['label'].value_counts()


label
negative    3241
positive    1808
neutral       72
Name: count, dtype: int64

## Class Distribution Analysis

The training dataset contains three sentiment classes:

| Sentiment | Count | Percentage |
|-----------|------:|-----------:|
| Negative | 3241 | 63.29% |
| Positive | 1808 | 35.30% |
| Neutral | 72 | 1.41% |

### Observation

The dataset is imbalanced. The Negative class dominates the dataset, while the Neutral class is significantly underrepresented.

### Potential Impact

An imbalanced dataset may bias machine learning models toward predicting the majority class. Therefore, evaluation metrics such as Precision, Recall, and F1-score will be used alongside Accuracy. Class balancing techniques will be considered later if model performance on the Neutral class is unsatisfactory.

In [12]:
# Final Validation Checks

train_df.isnull().sum()

tweet    0
label    0
dtype: int64

Observation:
The dataset contains no missing values in either the tweet or sentiment label columns. No missing-value treatment is required.

In [13]:
train_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5121 entries, 0 to 5120
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   tweet   5121 non-null   str  
 1   label   5121 non-null   str  
dtypes: str(2)
memory usage: 80.1 KB


Observation:
Both columns have the correct datatype (string). No datatype conversion is required.

In [15]:
train_df["tweet"].str.len().describe()

count    5121.000000
mean      116.007225
std        63.676509
min         9.000000
25%        66.000000
50%       102.000000
75%       155.000000
max       279.000000
Name: tweet, dtype: float64

Observation:
Tweet lengths range from 9 to 279 characters, with an average length of approximately 116 characters. This indicates a healthy distribution suitable for text classification.

In [16]:
train_df.sample(10)

,tweet,label
3956,shey be na why me n u still fit play under com...,positive
4389,momsi dey beg me make i stay house my sister d...,positive
4813,hello is this how mikel or the entire gonna br...,neutral
2679,the kind sweat wey i don sweat today ehnnn eve...,negative
657,e be like say this one dey ment you dey order ...,negative
1688,whats this nonsense we call traffic sef which ...,negative
384,i tried to clear small bills in our house toda...,negative
1189,please lets join hands and find that eagle in ...,negative
814,wetin him de see for ground u no de see for on...,negative
2184,so im hearing a mans voice from my neighbors r...,negative


Observation:
Random sampling confirmed that the tweets and corresponding sentiment labels appear correctly aligned. No corrupted records were observed.

# Validation Summary

## Dataset Overview

- Total Records: 5,121
- Features: 2
- Language: Nigerian Pidgin
- Sentiment Classes: Positive, Negative, Neutral


## Validation Results

### Missing Values

- No missing values were found.
- All records contain both tweet text and sentiment labels.

Status: PASS


### Duplicate Records

- 480 duplicate rows were identified.
- These duplicates will be removed during the preprocessing stage.

Status: REQUIRES CLEANING


### Data Types

- tweet → String
- label → String

Both columns have the correct data type.

Status: PASS


### Sentiment Distribution

| Sentiment | Count |
|-----------|------:|
| Negative | 3241 |
| Positive | 1808 |
| Neutral | 72 |

Observation:

The dataset is imbalanced, with the Negative class representing the majority of the data and the Neutral class being significantly underrepresented.

This imbalance will be considered during model evaluation.


### Tweet Length Analysis

- Minimum Length: 9 characters
- Average Length: 116 characters
- Median Length: 102 characters
- Maximum Length: 279 characters

Observation:

Tweet lengths fall within a reasonable range for sentiment analysis.


### Random Sample Inspection

Random inspection confirmed that:

- Tweets are readable.
- Labels are correctly assigned.
- No corrupted records were observed.

Status: PASS


# Conclusion

The dataset is well-structured and suitable for sentiment analysis.

Key findings include:

- No missing values.
- Correct data types.
- Presence of duplicate records.
- Imbalanced sentiment classes.
- Healthy tweet length distribution.

Overall, the dataset is ready to proceed to the preprocessing stage, where duplicate removal and text cleaning will be performed.